### ISSUE TITLE
Clean up of `USING SCHEMA FILE`

### ISSUE THREAD

**Issue by nmeagan11** [MaterializeInc/materialize#14911]
We need to do some tech debt clean up of `FROM SCHEMA FILE` which has been gated behind unsafe mode (see https://github.com/MaterializeInc/materialize/pull/13703) but has some internal tests which rely on it. I'm breaking this up into a separate issue for tracking in order to close out https://github.com/MaterializeInc/materialize/issues/12855.

**Comment by sploiselle** [comment]
Will work on this as a filler while waiting for design feedback on #14266

**Comment by sploiselle** [comment]
cc @benesch @philip-stoev: would love for your feedback/thoughts (or suggestions for someone else to review). If I don't hear back tomorrow I'll put up a draft PR with the proposal that adds a new command, as that's least invasive.

To solve this issue, we need to:
- Compile protobuf descriptor files
- Make the compiled files' contents referenceable in testdrive

# TL;DR
We can solve this by:
- Adding a new command that reads files into parmeters
- Doing protobuf compilation during parsing

# Challenges

## Testdrive's DSL + implementation
Testdrive has two dimensions that complicate this:
- All testdrive actions besides `set` (which sets a variable) don't execute until all statements in the file/context have been parsed, i.e. they are deferred.
- Testdrive eagerly inlines its special parameters during parsing (i.e. they cannot be set after parsing); this is necessary because of how `Action`s are handled––they must implement an `undo` method that trivially lets them be undone to reset the environment's state. This is done in reverse order of the actions in the file to simplify `DROP` operations with dependencies.

### Why the simple solution doesn't work

Because of the above, you cannot compile protobuf files and stick them in a variable using the current structure--once you leave the parsing stage, you cannot set variables.

You cannot easily create a second, dynamic variable step either, because the statements are run in reverse order, so the variables you want statements to access will not exist in scope yet. This is essentially intractable because you also cannot reasonably parse all of the variables before the reverse execution because variables may be overwritten and it's unclear which value was referred to in which state.

# Solutions

## Parameterize execution

We could create a new command like `load-file-into-param file=input.pb encoding=hex`, which loads the file's contents into `Vec<String>` on `State`, and then like so

```sql
$ protobuf-compile-descriptors inputs=schema.proto output=schema.pb

$ load-file-into-param file=input.pb encoding=hex

> CREATE SOURCE input_proto
  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-input_proto-${testdrive.seed}')
  FORMAT PROTOBUF MESSAGE '.Input' USING SCHEMA $1
```

**Downside** - Not very general, and generalizing this is hard/awkward, e.g `execute-with-params "action_0" "action_1"...`? The only action we really need right now is reading from files, but maybe this is more attractive than I'm giving it credit for.

## Execute proto compilation during parsing

`protobuf-compile-descriptors` could execute to completion during parsing, at which point its contents could be assigned to a variable, i.e. add a `var=foo` argument to it.

**Why?** Actions that do not have a meaningful `undo` operation can trivially be handled during parsing. The downside here is if the operation is expensive, it will delay finding parsing errors that will cause the test to fail before you materialized any benefit from having done the expensive work.

**Downside** This also means that `file-append` needs to execute during parsing, which breaks the semantics of `file-delete`. The one use of `file-delete` is in a protobuf file where we could instead use versions of files instead of the same file name, though, so I don't think this invalidates this change.

## Second parameter pass

We could add a second parameter pass `!!{param}`, which gets replaced  in each statement dynamically (i.e. adding a new set of parameters to `State` but:
- We would have to disallow assigning to the variable more than once
- The values could not be used in a way that could impact the `undo` operation

Populating these values could be done either on statements (`dynamic-var=foo`) or on files (`set-dynamic-from-file file=input.pb var=foo`)

**Downside** This seems dicey and most likely to introduce foot guns.

**Comment by benesch** [comment]
Remind me, too: Brennan had to write an mzcompose composition to test
something like SCHEMA EARLIEST—I forget exactly what—because of the same
limitation in testdrive. His thing would have been much simpler as a
testdrive script if testdrive supported runtime variable evaluation. We
should get that fixed up too as part of this work.

**Comment by sploiselle** [comment]
@brennanvincent Can you point me to what Nikhil mentioned here?

**Comment by benesch** [comment]
It's this PR: https://github.com/MaterializeInc/materialize/pull/12446

I can't find my branch on the moment. It might be on my Linux machine, but don't presently have access to it. Regardless, that branch didn't do a whole lot. It's mostly as simple as removing every `fn undo` in testdrive, and then adding a new Kafka cleanup function that runs at script start that nukes any topic whose name matches `testdrive-*`.

**Comment by sploiselle** [comment]
Bleh--parameterized execution doesn't work because we don't get the parameters until we have access to a session, which is too late for the instances where we need these values.

I'm instead inclined toward a second parameter pass at runtime. Think this can be handled with a visitor. If this impacts performance too severely, it's possible that we gate the checks with a command like `dynamic-vars` that permits replacement only in the next statement that runs.

**Comment by benesch** [comment]
Woohoo! Thanks, @sploiselle!

**Comment by benesch** [comment]
Oh, heck yes! Thank you for picking this one up, Sean.

I’m a bit sorry you had to go so deep on the proposals here—this problem
has been known to me for a while and I never got around to writing it down.
I do have yet another alternative to recommend though: ditch the evaluation
of set at parse time. It was a mistake. As you say, the problem is that
undo steps are run in reverse. Well, the whole undo thing never actually
worked out, and we have a bunch of cleanup that executes outside of the
redo/undo framework.

I have a stale branch that removes undo, in the service of moving the
evaluation of set to runtime. I’ll try to remember put that PR up tonight,
but I might forget. If I do, please remind me tomorrow.
